### PR TITLE
想像上を操作

### PR DESCRIPTION
**Pull request by myxyy**
## 概要

https://github.com/MLShukai/ami-vconf24/issues/149

### 既知の問題
* outputフォルダが出力される

## Submit前の確認項目

- [ ] タイトルは一目でわかるようにし、説明文は PR を簡潔に説明するようにしましたか?
- [ ] あなたの PR が、異なる変更を束ねたものではなく、ひとつのことだけを行うものであることを確認しましたか?
- [ ] この PR で導入されたすべての変更点をリストアップしましたか?
- [ ] PR を`make run`コマンドでローカルでテストしましたか?

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -23,6 +23,7 @@ mlagents-envs = {git = "https://github.com/MLShukai/ml-agents.git", subdirectory
 hydra-joblib-launcher = "^1.2.0"
 matplotlib = "^3.9.2"
 seaborn = "^0.13.2"
+keyboard = "^0.13.5"
```
Comment (Geson-anko) [inline pyproject.toml:26]: `keyboard`モジュールの依存関係に関しては `play_imagination_world.py`のスクリプトの冒頭に記述してください。プロジェクト自体に依存関係を追加し続けると`poetry install`が時間がかかって大変になるので。
```suggestion
```

File: scripts/play_imagination_world.py
Comment (Geson-anko) [inline scripts/play_imagination_world.py:1]: hydraやomegaconfは使わずに、ハイパラは全て決め打ちで `IJEPAEncoder`,  `IJEPALatentVisualizationDecoder`, `ForwardDynamicsActionReward`をインスタンス化してください。このスクリプトを汎用的に使いたい気持ちはありますが、かけられる工数的に書捨てのスクリプトコードのような実装でOKです。

hiddenの初期状態は `torch.zeros`で実装してください。
https://github.com/MLShukai/ami-vconf24/blob/main/configs/interaction/agent/curiosity_image_multi_step_imagination.yaml#L3-L9

File: scripts/play_imagination_world.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    decoder = hydra.utils.instantiate(cfg.models.i_jepa_target_visualization_decoder.model)
+    decoder.to(device)
+    decoder.load_state_dict(torch.load(decoder_parameter_file, map_location=device))
+
+    forward_dynamics_parameter_file = PROJECT_ROOT / "data/2024-09-17_04-47-37,195367.ckpt/forward_dynamics.pt"
+    forward_dynamics = hydra.utils.instantiate(cfg.models.forward_dynamics.model)
+    forward_dynamics.to(device)
+    forward_dynamics.load_state_dict(torch.load(forward_dynamics_parameter_file, map_location=device))
+
+    class KeyboardActionHandler:
```
Comment (Geson-anko) [inline scripts/play_imagination_world.py:32]: classは関数の中ではなく、スクリプトの Global Fieldに書いてください。

File: scripts/play_imagination_world.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    embedding = encoder(initial_observation).squeeze(0)
+    hidden = torch.zeros(cfg.models.forward_dynamics.model.core_model.depth, cfg.models.forward_dynamics.model.core_model.dim, device=device)
+
+    while True:
+        handler.update()
+        action = handler.get_action()
+        print(action)
+        next_embedding_dist, _, _, hidden = forward_dynamics(embedding, hidden, action)
+        embedding = next_embedding_dist.rsample()
+        reconstruction = decoder(embedding.unsqueeze(0))
```
Comment (Geson-anko) [inline scripts/play_imagination_world.py:76]: 0 - 255の間へmin max normalizationをお願いします。
```suggestion
        reconstruction = decoder(embedding.unsqueeze(0))
        reconstruction = (reconstruction - reconstruction.min()) / (reconstruction.max() - reconstruction.min() + 1e-8) * 255
```

File: scripts/play_imagination_world.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        cv2.imshow("imagination world", reconstruction.squeeze(0).permute(1,2,0).to("cpu").detach().numpy())
```
Comment (Geson-anko) [inline scripts/play_imagination_world.py:77]: RGB 2 BGRをお願いします。また uint 8型への変更もお願いします。

**Review by Geson-anko — CHANGES_REQUESTED** [review]
@myxyy 実装ありがとうございます。大体できてきましたね。Keyboardが取れない問題はRemoteからアクセスしている場合に発生するかもしれません。`Moonlight`とかなら大丈夫だった気がしますが、keyboardの動作確認に関してはこちらでも試してみます。

File: scripts/play_imagination_world.py [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+                self.actions["Jump"],
+                self.actions["Run"],
+            ]
+        )
+
+
+@hydra.main(
+    config_path="../data/2024-09-17_04-47-37,195367.ckpt", config_name="launch-configuration.yaml", version_base="1.3"
+)
+def main(cfg: DictConfig) -> None:
```
Comment (Geson-anko) [inline scripts/play_imagination_world.py:68]: 決め打ちでハイパラ書いていくのでhydraいらないと思います．
```suggestion
def main() -> None:
```

File: scripts/play_imagination_world.py [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+            [
+                self.actions["MoveVertical"],
+                self.actions["MoveHorizontal"],
+                self.actions["LookHorizontal"],
+                self.actions["Jump"],
+                self.actions["Run"],
+            ]
+        )
+
+
```
Comment (Geson-anko) [inline scripts/play_imagination_world.py:64]: パラメータのダウンロードリンクの記述お願いします．
```suggestion
# Parameter: https://drive.google.com/file/d/1saWq1fAbyA_BCZryUiWgPixJ6JCssLy0/view?usp=sharing
```

**Review by Geson-anko — CHANGES_REQUESTED** [review]
@myxyy hydraやOmegaconfは一旦無しで実装お願いします．
余裕が出たら汎用化しますので

File: scripts/play_imagination_world.py [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+        ),
+        action_hat_dist_head=DiscretePolicyHead(dim_in=2048, action_choices_per_category=ACTION_CHOICES_PER_CATEGORY),
+        reward_hat_dist_head=FullyConnectedFixedStdNormal(dim_in=2048, dim_out=1, squeeze_feature_dim=True),
+    )
+    forward_dynamics.to(device)
+    forward_dynamics.load_state_dict(torch.load(forward_dynamics_parameter_file, map_location=device))
+
+    handler = KeyboardActionHandler()
+
+    initial_observation = torch.randn(
```
[the remark this answers is not in the document]
Comment (myxyy) [reply scripts/play_imagination_world.py]: すみませんpush忘れでした
画像を読み込むようになってます

File: scripts/play_imagination_world.py (hunk 5 above)
Comment (myxyy) [reply scripts/play_imagination_world.py]: 対応しました

File: scripts/play_imagination_world.py (hunk 6 above)
Comment (myxyy) [reply scripts/play_imagination_world.py:65]: 対応しました

File: pyproject.toml (hunk 1 above)
Comment (Geson-anko) [reply pyproject.toml]: そうですね． コメントで `# pip install keyboard`と付けておきましょう．

**Review by Geson-anko — APPROVED** [review]
動作確認しました．回転には機敏に反応しますね．
若干修正したので，それを確認したらMergeしてください．
https://github.com/MLShukai/ami-vconf24/pull/152/commits/18a707f17965b792c359561e0fb7b71dccc96195
